pkg/controller/telemetry/operation_metric.go

```diff
@@ -0,0 +1,171 @@
+	ENABLE_ENCODING_METADATA: "ENABLE_ENCODING_METADATA",
+}
+
+type OperationMetricController struct {
+	operationMetricCache map[operationMetricLabels]*operationUsageInfo
+	mutex                sync.RWMutex
+}
+
+type operationUsageMetric struct {
+	startTime     uint64
+	endTime       uint64
+	pidTgid       uint64
+	operationType uint32
+}
+
+type operationUsageInfo struct {
+	durations     []uint64
+	pidTgid       uint64
+	operationType uint32
+}
+
+type operationMetricLabels struct {
+	nodeName      string
+	pidTgid       string
+	operationType string
+}
+
+func NewOperationMetric() *OperationMetricController {
+	return &OperationMetricController{
+		operationMetricCache: map[operationMetricLabels]*operationUsageInfo{},
+	}
+}
+
+func (m *OperationMetricController) Run(ctx context.Context, KmeshPerfInfo *ebpf.Map) {
+	if m == nil {
+		return
+	}
+	var err error
+	readerPerformance, err := ringbuf.NewReader(KmeshPerfInfo)
+	if err != nil {
+		log.Errorf("open performance notify ringbuf map FAILED, err: %v", err)
+		return
+	}
+	defer func() {
+		if err := readerPerformance.Close(); err != nil {
+			log.Errorf("ringbuf reader Close FAILED, err: %v", err)
+		}
+	}()
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				time.Sleep(operationMetricFlushInterval)
+				m.updatePrometheusMetric()
+			}
+		}
+	}()
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			operationData := operationUsageMetric{}
+			rec := ringbuf.Record{}
+			if err := readerPerformance.ReadInto(&rec); err != nil {
+				log.Errorf("ringbuf reader FAILED to read, err: %v", err)
+				continue
+			}
+			if len(rec.RawSample) != OPERATION_USAGE_DATA_LEN {
+				log.Errorf("wrong length %v of a msg, should be %v", len(rec.RawSample), OPERATION_USAGE_DATA_LEN)
+				continue
+			}
+			operationData.startTime = binary.LittleEndian.Uint64(rec.RawSample[0:8])
+			operationData.endTime = binary.LittleEndian.Uint64(rec.RawSample[8:16])
+			operationData.pidTgid = binary.LittleEndian.Uint64(rec.RawSample[16:24])
+			operationData.operationType = binary.LittleEndian.Uint32(rec.RawSample[24:28])
+			metricLabels := m.buildOperationMetric(&operationData)
+			operationInfo := operationUsageInfo{
+				durations:     []uint64{operationData.endTime - operationData.startTime},
+				pidTgid:       operationData.pidTgid,
+				operationType: operationData.operationType,
+			}
+			m.mutex.Lock()
+			m.updateOperationMetricCache(operationInfo, metricLabels)
+			m.mutex.Unlock()
+		}
+	}
+}
+
+// buildOperationMetric builds the operation metrics labels using actual pod info.
+func (m *OperationMetricController) buildOperationMetric(data *operationUsageMetric) operationMetricLabels {
+	labels := operationMetricLabels{}
+	nodeName := os.Getenv("NODE_NAME")
+	if nodeName == "" {
+		nodeName = "unknown-node-name"
+	}
+	labels.nodeName = nodeName
+	labels.operationType = operationTypeMap[data.operationType]
+	labels.pidTgid = fmt.Sprintf("%d", data.pidTgid)
```

Comment:
pure pid does not make sense, ideally we should have correlate with pod name/namespace

Reply:
So i suggest first leave it unset

Reply:
The ability to retrieve pidTgid is retained, but it is not added to the labels.